File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

vector<int> solution(string msg) {
    vector<int> answer;
    
    unordered_map<string, int> msg_map;
    for(int i = 0 ; i < 26 ; i++){
        char c = 'A' + i;
        string t;
        t.push_back(c);
        msg_map[t] = i + 1;
    }
    
    int next_idx = 27;
    int i = 0;
    
    while(i < msg.size()) {
        
        string w;
        int j = i;
        
        while(j < msg.size() && msg_map.find(msg.substr(i, j - i + 1)) != msg_map.end()){
            w = msg.substr(i, j - i + 1);
            j++;
        }
        
        answer.push_back(msg_map[w]);
        
        if(j < msg.size()){
            msg_map[msg.substr(i, j - i + 1)] = next_idx++;
        }
        
        i += w.size();
        
    }
    
    
    
    
    return answer;
}

File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.java
import java.util.*;
import java.util.stream.*;

class Solution {
    public int[] solution(String msg) {
        List<Integer> answer = new ArrayList<>();
        
        // 기본 단어 추가
        List<String> dic = new ArrayList<>();
        for(int i = 0 ; i < 26 ; i++){
            char cur = 'A';
            cur += i;
            dic.add("" + cur);
        }
        
        // 단어 순회
        for(int i = 0 ; i < msg.length(); i++){
        
            String tmp = "" + msg.charAt(i);
            while(i + 1 < msg.length() &&
                 dic.indexOf(tmp + msg.charAt(i + 1)) >= 0){
                tmp = tmp + msg.charAt(i + 1);
                i++;
            }
            
            answer.add(dic.indexOf(tmp) + 1);
            
            if(i < msg.length() - 1){
                String next =tmp + msg.charAt(i + 1);
                if(!dic.contains(next))
                    dic.add(next);
            }
            
        }
        
        
        return answer.stream().mapToInt(Integer::valueOf)
            .toArray();
    }
}